src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java

```diff
@@ -0,0 +1,527 @@
+                  return new RegularFileStatus(digest.getSizeBytes(), DigestUtil.toBinaryDigest(digest));
+                }
+                if (fileStatus.hasDirectory()) {
+                  throw new RuntimeException("DIRECTORY");
+                }
+                if (fileStatus.hasSymlink()) {
+                  return new SymlinkFileStatus();
+                }
+                if (fileStatus.hasExternal()) {
+                  throw new RuntimeException("EXTERNAL");
+                }
+                throw new RuntimeException("SOMETHING ELSE");
+              }));
+    }
+  }
+
+  @Override
+  public BatchStat getBatchStatter() {
+    return new GrpcBatchStat();
+  }
+
+  @Override
+  public boolean canCreateSymlinkTree() {
+    return true;
+  }
+
+  @Override
+  public void createSymlinkTree(Map<PathFragment, PathFragment> symlinks, PathFragment symlinkTreeRoot)
+      throws ExecException, InterruptedException {
+    // The provided set of symbolic links may be too large to provide to
+    // the remote output service at once. Partition the symbolic links
+    // in groups of 1000, so that BatchCreateRequest messages remain
+    // small enough.
+    UnmodifiableIterator<List<Map.Entry<PathFragment, PathFragment>>> symlinkBatchIterator =
+        partition(
+            filter(symlinks.entrySet().iterator(), (symlink) -> symlink.getValue() != null),
+            1000);
+    boolean cleanPathPrefix = true;
+    while (symlinkBatchIterator.hasNext()) {
+      List<Map.Entry<PathFragment, PathFragment>> symlinksBatch = symlinkBatchIterator.next();
+      BatchCreateRequest.Builder builder = BatchCreateRequest.newBuilder();
+      builder.setBuildId(stateAcrossBuilds.currentBuildId.toString());
+      builder.setPathPrefix(symlinkTreeRoot.toString());
+      builder.setCleanPathPrefix(cleanPathPrefix);
+      for (Map.Entry<PathFragment, PathFragment> symlink : symlinksBatch) {
+        OutputSymlink.Builder symlinkBuilder = builder.addSymlinksBuilder();
+        symlinkBuilder.setPath(symlink.getKey().toString());
+        symlinkBuilder.setTarget(symlink.getValue().toString());
+      }
+      blockingStub.batchCreate(builder.build());
+      cleanPathPrefix = false;
+    }
+  }
+
+  @Override
+  public void clean() throws ExecException, InterruptedException {
+    CleanRequest.Builder builder = CleanRequest.newBuilder();
+    builder.setOutputBaseId(outputBaseId);
+    blockingStub.clean(builder.build());
+  }
+
+  @Override
+  public ActionFileSystemType actionFileSystemType() {
+    return ActionFileSystemType.STAGE_REMOTE_FILES;
+  }
+
+  private class GrpcFileSystem extends DelegateFileSystem {
+    private final PathFragment originalOutputPath;
+
+    public GrpcFileSystem(FileSystem sourceDelegate, PathFragment originalOutputPath) {
+      super(sourceDelegate);
+      this.originalOutputPath = originalOutputPath;
+    }
+
+    @Override
+    protected byte[] getFastDigest(PathFragment path) throws IOException {
```

Comment:
Is this necessary for the solution to work at all or is it just an optimization to avoid calls going through FUSE? I saw in the doc, you mentioned that xattrs may be problematic for certain NFS implementations -- I guess this may heavily depend on where we would use NFS for this and what the capabilities are. Would you mind expanding on this a little?

Reply:
Sure thing! So there are multiple downsides to using extended attributes for this:

- Linux: As soon as your FUSE file system starts to announce the availability of the GETXATTR operation, it basically opens the floodgates for the kernel to call into this operation whenever it wants to do certain ACL-based permission checks. From a performance perspective, it's generally smarter to let GETXATTR return ENOSYS, and let Bazel send selective RPCs whenever it's desired to get the digest.

- macOS: Since this PR was initially published, bb\_clientd has gained support for NFSv4. The bb\_clientd virtual file system layer can now be exposed both over FUSE and NFSv4. Extended attributes are still problematic on NFSv4, as macOS only permits storing those in AppleDouble files (`._MyFile`) for the time being. The exact encoding of those files is not a public API, so we should not try to handcraft those.

- In general: when we use gRPC, we also have the ability to do fancy things like batching, which significantly reduces the number of context switches between Bazel and bb\_clientd. This is currently not implemented, but relying on extended attributes to convey this information already closes the door on that.